### ISSUE TITLE
CSS class-based replace

### ISSUE THREAD

**Issue by firedev** [rails/turbolinks#616]
Hi,

After working a bit with `Cells` and `Trailblazer` I have adopted some concepts from React and invented the view-less system where updates are happening from the controller, it looks a bit like Turbolinks3:

```
class CalendarController
    def update
     ...
      update_cells(
        concept('worktime/form', @worktime),
        concept('calendar/day', day),
        concept('calendar/stat', day.month, user: current_user)
      )
    end
end
```
`update_cells` just calls `update` method on each `concept` and glues the js string together:
```
class ApplicationController
  def update_cells(*cells)
    render text: cells.flatten.join
  end
end
```
Each cell has it's unique-ish document identifier, but instead of `id` I use unique CSS `class`. 

Why?

That lets me do sweeping updates on the page, so when one cell is changed, it is being re-rendered everywhere in the document. And the only thing that stops me from jumping on Turbolinks3 train is that Turbolinks uses `id`.

Is there a way to replace document fragments based on CSS `class`? Was this considered even? Why not?

Thanks.

**Comment by tortuetorche** [comment]
Hi @firedev,

In the Turbolinks docs, there is a concept of _namespaced_ ids, maybe you can use it instead of CSS class?

> If the id contains a colon, the key before the colon can also be targeted to replace many nodes with a similar prefix.

```html
<div id="comments:123"></div>
```

Cheers,
Tortue Torche

**Comment by firedev** [comment]
Yep, that is the problem. I don't have the luxury of React that keeps track of it's components and can't do cycle in rails for obvious reasons. All in all `id`-based approach seems like a dead-end to me. I use `unique` identifiers made from cell name and `dom_id` of the main object so it looks kind of like this:

`<div class="calendar-day_1234">`

This way I don't have to worry about updating the wrong fragment and if there are more than one - no problem.

**Comment by tortuetorche** [comment]
Sorry, I don't really understand what you said.
My English knowledge isn't as good as I hope.

If you use `dom_id()` helper, you can prefix it with `comments:`.

In a Rails console:
```ruby
comment = Comment.first
helper.dom_id(comment)
=> "comment_2"
helper.dom_id(comment, 'comments:')
=> "comments:_comment_2"
```
The generated `id` is a bit weird, but it should work.

Then when you do a:

```javascript
Turbolinks.replace('some HTML tags', {change: 'comments'});
```

Turbolinks will replace all the HTML nodes with an id beginning with this string: `comments:`.

**Comment by firedev** [comment]
I render the same `cell` in a few places on the page, I can't generate `#comments:comment_1:1` `#comments:comment_1:2` `#comments:comment_1:3`. But if using css selectors, or even `data-id` one is good enough: `.comments-comment_1`.

**Comment by tortuetorche** [comment]
Why do you render the same fragment in a few places of the page?

**Comment by firedev** [comment]
I am building a worktime calendar app, there are few stats cells on each end, there are day cells you can edit in a modal window and they are being updated on-the-fly.

Just remembers that this came up before and I was asking on StackOverflow: http://stackoverflow.com/questions/25418379/storing-object-id-in-html-document-id-vs-data-id-vs

Apparently there are cases when you want more than one kind of the same object on the page.

**Comment by tortuetorche** [comment]
Well, maybe you can do something like that:

```ruby
class Message < ActiveRecord::Base
  validates :title, presence: true
  attr_accessor :context_id
  
  def to_key
    super << @context_id
  end
end

**Comment by tortuetorche** [comment]
Or override the [`dom_id()`](http://api.rubyonrails.org/classes/ActionView/RecordIdentifier.html#method-i-dom_id) helper to add a third argument, who appends a suffix.

```ruby
comment = Comment.first
dom_id(comment, 'comments:', 1)
=> "comments:_comment_2_1"
```

**Comment by firedev** [comment]
I'll just rename the title to a more general 'selector-based' replacement, so it allows more freedom of speech.
